src/mtev_http.c

```diff
@@ -1168,15 +1229,165 @@ mtev_http_session_req_consume_chunked(mtev_http_session_ctx *ctx,
+      *out = NULL;
+      errno = -errno;
+      return -1;
+    }
+    o->size += out_size;
+    total_decompressed_size += o->size;
+    in->size -= in_size;
+    in->start += in_size;
+  }
+  (*last_out)->next = NULL;
+  return total_decompressed_size;
+}
+
+/* same as previous but gzip data */
+static ssize_t
+mtev_http_session_decompress_gzip(struct bchain *in, mtev_http_session_ctx *ctx, 
+                                  struct bchain **out, struct bchain **last_out)
+{
+  ssize_t total_decompressed_size = 0;
+  z_streamp zstream = ctx->req.zlib_decompress_ctx;
+  if (zstream == NULL) {
+    zstream = calloc(1, sizeof(z_stream));              
+    ctx->req.zlib_decompress_ctx = zstream;
+
+    int err = inflateInit2(ctx->req.zlib_decompress_ctx, 15 + 16);
+    if (err != Z_OK) {
+      mtevL(mtev_error, "gzip error initing the zstream: %s\n",
+            zstream->msg);
+      free(zstream);
+      ctx->req.zlib_decompress_ctx = NULL;
+      errno = -errno;
+      return -1;
+    }
+  } 
+
+  if (*out == NULL) {
+    *out = ALLOC_BCHAIN(DEFAULT_BCHAINSIZE);
+  }
+  struct bchain *o = *out;
+  *last_out = *out;
+
+  zstream->next_in = (Byte *)(in->buff + in->start);
+  zstream->avail_in = in->size;
+
+  while (zstream->avail_in) {
+    size_t out_size = o->allocd - o->size;
+    size_t zstream_total_out_pre_inflate = zstream->total_out;
+    size_t zstream_total_in_pre_inflate = zstream->total_in;
+    if (out_size == 0) {
+      struct bchain *temp = o;
+      o = ALLOC_BCHAIN(DEFAULT_BCHAINSIZE);
+      temp->next = o;
+      *last_out = o;
+      out_size = o->allocd;
+    }
+    zstream->next_out = (Byte *)(o->buff + o->size);
+    zstream->avail_out = o->allocd - o->size;
+    int x = inflate(zstream, Z_NO_FLUSH);
+    if (x != Z_OK && x != Z_STREAM_END) {
+      mtevL(mtev_error, "zlib error decompressing: %s\n",
+            ctx->req.zlib_decompress_ctx->msg);
+      inflateEnd(zstream);
+      free(zstream);
+      ctx->req.zlib_decompress_ctx = NULL;
+      RELEASE_BCHAIN(*out);
+      *out = NULL;
+      errno = -errno;
+      return -1;
+    }              
+    /* we only want the incremental size here */
+    o->size += zstream->total_out - zstream_total_out_pre_inflate;
+    in->start += zstream->total_in - zstream_total_in_pre_inflate;
+    in->size -= zstream->total_in - zstream_total_in_pre_inflate;
+    total_decompressed_size += o->size;
+  }
+
+  (*last_out)->next = NULL;
+  return total_decompressed_size;
+}
+
 int
 mtev_http_session_req_consume(mtev_http_session_ctx *ctx,
-                              void *buf, size_t len, size_t blen,
-                              int *mask) {
+                              void *buf, size_t user_len, size_t blen,
+                              int *mask) 
+{
   size_t bytes_read = 0,
-         expected = ctx->req.content_length - ctx->req.content_length_read;
+    expected = ctx->req.content_length - ctx->req.content_length_read,
+    wire_len = expected;
```

Comment:
my spidey-sense is tingling when I see two comma-separated variable definitions where execution order matters... I'm not sure C specifies that `expected = ...` will evaluate before `wire_len = expected`?